util/iputil/parse.go

```diff
@@ -0,0 +1,28 @@
+package iputil
+
+import (
+	"net"
+	"strings"
+)
+
+// IPVersion is the numerical version of the IP address spec (4 or 6).
+type IPVersion int
+
+// IP address versions.
+const (
+	IPvUnknown IPVersion = 0
+	IPv4       IPVersion = 4
+	IPv6       IPVersion = 6
+)
+
+// ParseIP parses v as an ip address returning the result and version, or nil and unknown if invalid.
+func ParseIP(v string) (net.IP, IPVersion) {
+	if ip := net.ParseIP(v); ip != nil {
+		if strings.ContainsRune(v, ':') {
```

Comment:
net.ParseIP automatically converts IPv4 addresses to IPv6. There's no way to tell if the original value was IPv4 or IPv6 without the above check. This approach is common in other Go libraries.